### ISSUE TITLE
[Refactor]: Improve typing and sanitize docs

### ISSUE THREAD

**Issue by Borda** [roboflow/rf-detr#564]
### Search before asking

- [x] I have searched the RF-DETR [issues](https://github.com/roboflow/rf-detr/issues) and found no similar feature requests.

### Description

I suggest we remove them from the docstrings and rely solely on the signature. Here is the breakdown of why the Python ecosystem has shifted this way:

### 1. The Modern Approach (Signature as Truth)
Since PEP 484, the standard is to treat the function signature as the **Single Source of Truth**.
Tools like **mypy**, **Pyright**, and our IDEs rely entirely on the signature to validate our code. By keeping the focus there, we ensure that the types we document are the actual types being enforced by the interpreter.

### 2. The Legacy Approach (Why to avoid duplication)
Writing types in the docstring was necessary before Python 3.5, but today it is considered a violation of **DRY (Don't Repeat Yourself)**.
* **Maintenance Burden:** It creates a "split truth." If we change a variable from `int` to `float` in the code but forget to update the English text in the docstring, our documentation becomes factually incorrect.
* **Style Standards:** The **Google Python Style Guide** now explicitly states: *"The description should include the type if and only if the function does not have a type annotation."*

### 3. Do we need both for the docs?
No. You might be worried that removing them will break our generated documentation, but tools like **Sphinx** (via `sphinx-autodoc-typehints`) or **MkDocs** now automatically pull the types from the signature and insert them into the generated HTML. We get the same result without the manual effort.

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by Abdul-Mukit** [comment]
Are there any more scripts that we can take on?
I saw the migration plan to PyTorch Lightning. I was wondering if the MRs related to improving typing in existing scripts are still needed.

**Comment by farukalamai** [comment]
Hi @Borda ! I've opened a PR for the rfdetr/evaluation and rfdetr/utilities modules: #888

**Comment by hinogi** [comment]
how strict should the typing be? I experienced that pylance and such will 'wrongly' narrow types down to Unknown although they have been in instance check before.
Do you want an explicit cast in that case so the type checker is happy?

**Comment by hinogi** [comment]
Well, I am not a python dev, but I know my way around stuff. There are a lot of open traps being laid out in the source code for unexperienced developers I can tell.
It is pretty messy.

**Comment by Jonathan-Jesni** [comment]
Hi @Borda 
I'm new to open source and would love to help with this, is there a module I can pick up to add typing to?

**Comment by Borda** [comment]
> I'm new to open source and would love to help with this, is there a module I can pick up to add typing to?

Choose a small one, for typing, check if there are any type ignore in code or ignored modules in pyproject.toml

**Comment by Jonathan-Jesni** [comment]
Thanks @Borda I'll take "rfdetr.training.drop_schedule" then

**Comment by hinogi** [comment]
@Borda Would you mind to post some instructions on what to do first before starting to work to check if everything is working as expected so that the unexperienced does not get ambushed by error messages and missing stuff?

**Comment by hinogi** [comment]
Just some things I wasn't aware of.
VSCode's default tooling is pylance for python. This project uses mypy.
For gradual implementation of mypy typings they are on an ignore list which I wasn't aware of because pylance was already reporting errors.
First step should be to remove them from the ignore list in the project and also install mypy for vscode.

**Comment by hinogi** [comment]
https://github.com/roboflow/rf-detr/blob/develop/src/rfdetr/training/callbacks/coco_eval.py#L62-L81

@Borda Should I fix this also with direct access and ignore?

**Comment by atikulmunna** [comment]
Working through the rest of the `ignore_errors` block and ran into something worth settling once instead of four times.

Of the 76 errors left, 26 are the same one: `Missing type parameters for generic type "ndarray"`. They are spread across four modules:

| module | count |
|---|---|
| `rfdetr.datasets.yolo` | 16 |
| `rfdetr.datasets.transforms` | 5 |
| `rfdetr.datasets.synthetic` | 4 |
| `rfdetr.datasets.save_grids` | 1 |

The codebase already has two conventions for this, both in code that passes strict today:

- `NDArray[np.float32]` (via `from numpy.typing import NDArray`), 28 uses across `export/`, `models/matcher.py`, and `training/callbacks/drop_schedule.py`
- `np.ndarray[Any, Any]`, 7 uses in `evaluation/coco_eval.py` and `evaluation/keypoint_oks.py`

Which one do you want in the dataset modules? Whoever gets to `yolo.py` first effectively decides for the other three, so I would rather ask than guess and have it reworked later.

For what it is worth, I lean toward the precise form here, because in these modules the dtype is not ambiguous. `yolo.py` sets it explicitly at every construction site: `np.float32` for boxes, polygons, and keypoints, `np.int64` for `class_id`, `bool` for masks. `np.ndarray[Any, Any]` seems like the right escape hatch for `evaluation/`, where the arrays come back out of `faster_coco_eval`, but in the dataset code it would quiet mypy without saying anything true.

Happy to take `yolo.py` on whichever convention you pick. Picking up `rfdetr.detr` in the meantime since it does not touch this question.

**Comment by Borda** [comment]
> Happy to take `yolo.py` on whichever convention you pick. Picking up `rfdetr.detr` in the meantime since it does not touch this question.

@atikulmunna, please go ahead!

**Comment by Jonathan-Jesni** [comment]
Update from my side: I had originally picked up `rfdetr.training.drop_schedule`, but that was completed in #1172 before I got to it, so I moved on to a module that was still unclaimed.

I've opened #1304 for `rfdetr.datasets.save_grids`.

Removing it from the `ignore_errors` list surfaced two `type-arg` errors, both fixed by adding type parameters:

- `DatasetGridSaver.__init__`: `DataLoader` becomes `DataLoader[Any]`, matching `RFDETRDataModule.train_dataloader()` and `val_dataloader()` in `module_data.py`, which are the only call sites
- `DatasetGridSaver._extract_masks`: `np.ndarray` becomes `np.ndarray[Any, Any]`, matching the style used in `coco_eval`, `export.main`, and `detr`

No runtime changes, since the module uses `from __future__ import annotations`. `mypy src/rfdetr/` is clean for the module and the existing tests pass unmodified.

One question for maintainers: `_extract_masks` always returns a boolean array, so `np.ndarray[Any, np.dtype[np.bool_]]` would be more precise and also type-checks cleanly, matching the pattern in `rfdetr.evaluation.matching`. I went with `np.ndarray[Any, Any]` as it's the more common convention across the codebase, but happy to tighten it if you'd prefer the specific annotation.

**Comment by adhavan18** [comment]
@Borda if it's free, I'd like to take **`rfdetr.datasets.synthetic`** from the `ignore_errors` list.

checking first because the list is nearly done and I'd rather not collide with anyone. as far as I can tell five modules are left:

```toml
{ module = [
  "rfdetr",
  "rfdetr.config",
  "rfdetr.datasets.coco",
  "rfdetr.datasets.synthetic",
  "rfdetr.datasets.transforms",
], ignore_errors = true },
```

and I couldn't find an open PR claiming any of them — #1337 and #1338 are the most recent and both merged. `synthetic` looks like the smallest of the five at ~516 lines and it doesn't touch the public API, so it seemed the sensible one to start on rather than `config` or `coco`.

on the `ndarray` generics question @atikulmunna raised: I'll follow whatever convention landed in the modules already done rather than introducing a different one, and if `synthetic` turns out to need a call that hasn't been made yet I'll ask here instead of picking unilaterally.

happy to take a different module if this one is spoken for.

**Comment by Jonathan-Jesni** [comment]
Picking up `rfdetr.config` from the remaining `ignore_errors` list. Checked and don't see an open PR touching it. Will link a PR here once ready, happy to switch if it's already spoken for.

**Comment by Jonathan-Jesni** [comment]
I've opened #1362 for `rfdetr.config`.

Removing it from the `ignore_errors` list surfaced three errors, none of them the missing-type-parameter pattern from `atikulmunna`'s note above:

- `_detect_device()`: a real gap, `accelerator.is_available()` was called without narrowing `accelerator` itself to non-`None` (only the separately-derived `current_accelerator` was checked). Added the missing guard.
- `RFDETRLargeConfig.projector_scale`: typed narrower than the base class (`list[Literal["P4",]]` vs `ModelConfig`'s `list[Literal["P3", "P4", "P5"]]`), a real substitutability violation since `list` is invariant. Widened to match the base class and sibling configs.
- `TrainConfig._serialize_augmentation_backend()`: returned `Enum.value` (typed `Any` in typeshed) from a function declared `-> str`. Wrapped in `str(...)`.

`mypy src/rfdetr/` is clean for the module and `pytest tests/ -k config` passes unmodified (579 passed).

**Comment by Jonathan-Jesni** [comment]
Picking up `rfdetr.datasets.transforms` next. Checked and it doesn't look like anyone's on it right now, will drop the PR link here once it's up.

**Comment by arubittu** [comment]
Picking up `rfdetr.datasets.coco`, the last module in the `ignore_errors` list. I checked the current open PRs and do not see one touching this file. I will keep this limited to strict mypy coverage and behavior-neutral typing fixes, then link the PR here.

**Comment by Borda** [comment]
Fully complete, great effort, thank you everyone!
